sys/kern/proc.c

```diff
@@ -569,42 +569,81 @@ __noreturn void proc_exit(int exitstatus) {
   thread_exit();
 }
 
-int proc_sendsig(pid_t pid, signo_t sig) {
-
-  proc_t *target;
-
-  if (pid > 0) {
-    WITH_MTX_LOCK (all_proc_mtx)
-      target = proc_find(pid);
-    if (target == NULL)
-      return EINVAL;
-    sig_kill(target, sig);
-    proc_unlock(target);
-    return 0;
-  }
-
-  /* TODO send sig to every process for which the calling process has
-   * permission to send signals, except init process */
-  if (pid == -1)
-    return ENOTSUP;
-
+static int proc_pgsignal(proc_t *p, cred_t *cred, pgid_t pgid, signo_t sig) {
   pgrp_t *pgrp = NULL;
 
   WITH_MTX_LOCK (all_proc_mtx) {
-    if (pid == 0)
+    if (pgid == 0) {
       pgrp = proc_self()->p_pgrp;
-
-    if (pid < -1) {
-      pgrp = pgrp_lookup(-pid);
+    } else {
+      pgrp = pgrp_lookup(pgid);
       if (!pgrp)
-        return EINVAL;
+        return ESRCH;
     }
     mtx_lock(&pgrp->pg_lock);
   }
 
-  sig_pgkill(pgrp, sig);
+  proc_t *target;
+  int send = 0, error = 0;
+  TAILQ_FOREACH (target, &pgrp->pg_members, p_pglist) {
+    WITH_PROC_LOCK(target) {
+      if (!(error = proc_cansignal(p, cred, target, sig))) {
+        sig_kill(target, sig);
+        send++;
+      }
+    }
+  }
   mtx_unlock(&pgrp->pg_lock);
-  return 0;
+
+  /* We return error when signal can't be send to any process. Returned error is
+   * last error obtained from checking privileges.*/
+  return send > 0 ? 0 : error;
+}
+
+int proc_sendsig(proc_t *p, pid_t pid, signo_t sig) {
+
+  if (sig >= NSIG)
+    return EINVAL;
+
+  int error = 0;
+  proc_t *target;
+  cred_t *cred = kmalloc(M_TEMP, sizeof(cred_t), 0);
+
+  if (!cred)
+    return ENOMEM;
+
+  WITH_PROC_LOCK(p) {
+    cred_copy(cred, p);
+  }
```

Comment:
Would it be true if we will have multiple threads in the future? Then another thread can change credentials in the same time. I am not sure if I can leave it without locking (it would be hard to find when aligning code to multiple threads).

Reply:
I wouldn't worry about multithreaded processes for now.